File: 1-Search-Algorithms/ucs.py
# Uniform-Cost-Search Algorithm
from heapq import heappush, heappop

def ucs(g, node_types):
    pq = [(0, 0)]
    path = []
    while True:
        cost, node = heappop(pq)
        path.append(node)
        if node_types[node] == "goal":
            break
        if not g[node]:
            path.pop()
            continue
        for nd, cs in g[node]:
            heappush(pq, (cost+cs, nd))
    return path, cost


graph = {
    0: [(1, 5), (2, 2)],
    1: [(2, 1)],
    2: [(3, 7)],
    3: []
}

node_type = {
    0: "start",
    1: "normal",
    2: "goal",
    3: "normal"
}

print(ucs(graph, node_type))

File: 1-Search-Algorithms/ucs.clj
; Uniform-Cost-Search Algorithm

(defn ucs [graph]
  (loop [priority-queue [[[:S] 0]]] ;; cada elemento é [caminho custo]
    (let [[[path cost] & rest-pq] (sort-by second priority-queue)
          node (last path)
          neighbors (get graph node)]
      (if (= node :G)
        path
        (let [new-nodes (map (fn [[nd cs]] [(conj path nd) (+ cost cs)]) neighbors)]
          (recur (into rest-pq new-nodes)))))))

(def problem {:S [[:A 5] [:B 2]]
              :A [[:B 1]]
              :B [[:G 7]]
              :G []})

(println (ucs problem)) 

